### PR TITLE
[REG-1706] Playback for programmatic playback of recordings

### PR DESCRIPTION
**Pull request by vontell**
Adds documentation for the ability to playback a recording from a Unity Test Runner file.

Loom Video: https://www.loom.com/share/dae3882584f042f5a2db866a5253f862?sid=9bcfdc9e-47c8-4981-9451-7633d3a4a5d9
BossRoom PR: https://github.com/Regression-Games/RGBossRoom/pull/57
SDK Changes PR: https://github.com/Regression-Games/RGUnityBots/pull/183

Screenshot 2024-05-29 at 11 40 56 AM
Screenshot 2024-05-29 at 11 41 03 AM
Screenshot 2024-05-29 at 11 41 10 AM

### CODE REVIEW

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (svolokh) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:255]: Non-blocking: See my suggestion about generalizing the callback parameter here: https://github.com/Regression-Games/RGUnityBots/pull/183#discussion_r1619247539

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (svolokh) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:60]: Blocking: As a developer, I think having `LogAssert.ignoreFailingMessages` be a requirement for using this testing library would be a deal breaker for me. There are very few generic test oracles for games, and error logs happen to be one of them. I definitely would want my CI/CD pipeline to tell me if an unexpected error was encountered during the testing session.

From the Loom video I gathered that the motivation for this was due to an error coming out of the SDK about screenshots. If this is a non-critical error, then I would recommend to change that specific error to a warning, rather than having the test case ignoring all errors coming out of Unity entirely.

**Review by svolokh — CHANGES_REQUESTED** [review]
Everything mostly looks good to me, I have one blocking item below about the usage example that I think needs discussion.

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:46]: This path isn't correct

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:60]: Agree this shouldn't be turned on

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:235]: Non-blocking I guess, but a re-occurring discussion point in doc reviews

We really shouldn't be documenting APIs manually on the doc site.

If we want to create an automatically generated API docs site from the build that sounds fine/good, but if I update the code, I highly doubt I'll remember to come retype this doc page.

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (abeizer) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:39]:
```suggestion
A recorded Gameplay Session can be played back programmatically from a 
```

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (abeizer) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:41]:
```suggestion
Create a Play Mode test file and use the `RGTestUtils` class to load and play a recording. This 
```

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (abeizer) [inline docs/gameplay-sessions/gameplay-sessions-usage.mdx:42]:
```suggestion
will also record the replay as a new Gameplay Session.
```

**Review by abeizer — APPROVED** [review]
Some rewording suggestions.

Not blocking for this task, but something to consider: Checking in recordings as part of the project would become unmanageable very quickly. We should think about querying a user's GameplaySessions from RGService and having them select one from the in-game overlay instead of their file system. It will likely make the process feel more cohesive and will set us up to be able to share GameplaySessions via RG accounts rather than as project assets. I suppose if someone isn't logged in through the SDK then file system is fine, but if we're syncing data then we should take advantage of that.

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:46]: Fixed here now and in the other locations where I copied this from. Since the new definition is quite long, I chose to define it in one spot and then reference it in other locations

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:60]: Ok so - I spent about 2 hours really digging into this, and as explained on the call, there are a couple issues that make this pretty tough to solve:

- The exception is not actually thrown; I believe under the hood, they catch this and do `Debug.LogError`
- I can't disable error logging for just this line, because it spawns a thread, and so there is no guarantee when this will run.
- There is no config for disabling certain packages or logs from being printed. I _could_ implement my own logger, but I found that it's tough to derive or chain this logger to whatever logger the project may be using, and I don't want to override their logger in the test temporarily (i.e. the approach requires overriding / setting the Debug.unityLogger field)

Because this is only relevant in Mac, I have decided to leave this, but comment it out, and leave a comment for mac users about this error. If this is ok with everyone, then awesome! Otherwise, I will spend time implementing a custom logger

Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: Why not just wrap it in something like this

```csharp
#if UNITY_EDITOR_OSX || UNITY_RUNTIME_OSX
#endif 
```

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:23]: this line is misleading... `referenceSessionId` simply doesn't exist for 'recording'

**Review by RG-nAmKcAz — APPROVED** [review]
Approving so i don't have to read this again, but...

blocking comment on the `referenceSessionId` should be corrected before merge

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: https://docs.unity3d.com/Manual/PlatformDependentCompilation.html

**Review by svolokh — APPROVED** [review]
Approving since if we don't have a solution on hand for the ignoreFailingAsserts issue then I think we can move forward for now. I do think that ideally we should eventually find a way to avoid this requirement, since it hampers the effectiveness of the tests.

However, as a parting thought as a possible solution for this, perhaps temporarily overriding the `Debug.unityLogger` field to reduce the severity of the error to a warning could be fairly easy/safe? I've written some pseudo-code for how that could work here: https://gist.github.com/svolokh/9956459e7961fda43cd4e262df31f745

File: docs/gameplay-sessions/gameplay-sessions-usage.mdx
Comment (svolokh) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx]: @vontell I've approved this for now since if we don't have a solution on hand I think it's OK to move forward with this for now (but I still think eventually we should find a way to reduce this requirement).

As I've written in my comment below, perhaps this approach to overriding the Debug.unityLogger could be a safe/easy way to do what is needed?  https://gist.github.com/svolokh/9956459e7961fda43cd4e262df31f745

Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-usage.mdx:70]: Thanks for the assistance @svolokh and @RG-nAmKcAz! I've added Zack's recommendation for now, but I also am glad you created a gist with an example - this is something I will look into! I'll make a note of it and see if I can do that at some point.

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: @RG-nAmKcAz I've went and updated this to hopefully explain in detail what this look like.

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (RG-nAmKcAz) [inline docs/gameplay-sessions/gameplay-sessions-reference.mdx:19]: What I was trying to get at Aaron is that this isn't correct.

its either `..._{currentSessionId}` or `...{currentSessionId}_{referenceSessionId}`  ; we don't show the currentSessionId 2x on recordings

**Review by RG-nAmKcAz — CHANGES_REQUESTED** [review]

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (vontell) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx:19]: Yes this will be fixed soon! I have no idea why, but I swear in one of my recordings, it had it twice... was it like that before? I deleted all of my recordings recently and now don't see that behavior. But yes that will be fixed soon! there are just some merge conflicts I need to address

**Review by addisonbgross — APPROVED** [review]

File: docs/gameplay-sessions/gameplay-sessions-reference.mdx
Comment (RG-nAmKcAz) [reply docs/gameplay-sessions/gameplay-sessions-reference.mdx]: @vontell  Was it intentional that your merge wiped out these changes ?

image

**Review by RG-nAmKcAz — APPROVED** [review]

**Pull request merged**
